contracts/crafting/facets/CraftingFacet.sol

```diff
@@ -0,0 +1,152 @@
+import "@openzeppelin-contracts/contracts/token/ERC721/extensions/ERC721Burnable.sol";
+import "@openzeppelin-contracts/contracts/token/ERC1155/utils/ERC1155Holder.sol";
+import "@openzeppelin-contracts/contracts/token/ERC721/utils/ERC721Holder.sol";
+import {MockTerminus} from "../../mock/MockTerminus.sol";
+import {MockErc20} from "../../mock/MockErc20.sol";
+import "../libraries/LibCrafting.sol";
+import "../../diamond/libraries/LibDiamond.sol";
+import "../../diamond/security/DiamondReentrancyGuard.sol";
+
+contract CraftingFacet is ERC1155Holder, ERC721Holder, DiamondReentrancyGuard {
+    function addRecipe(Recipe calldata recipe) public {
+        LibCrafting.CraftingStorage storage cs = LibCrafting.craftingStorage();
+        cs.numRecipes++;
+        cs.recipes[cs.numRecipes] = recipe;
+    }
+
+    function getRecipe(uint256 recipeId) public view returns (Recipe memory) {
+        return LibCrafting.craftingStorage().recipes[recipeId];
+    }
+
+    function numRecipes() public view returns (uint256) {
+        return LibCrafting.craftingStorage().numRecipes;
+    }
+
+    function craft(uint256 recipeId) public diamondNonReentrant {
+        Recipe memory recipe = LibCrafting.craftingStorage().recipes[recipeId];
+        for (uint256 i = 0; i < recipe.inputs.length; i++) {
+            if (recipe.inputs[i].tokenType == LibCrafting.ERC20_TOKEN_TYPE) {
+                ERC20Burnable token = ERC20Burnable(
+                    recipe.inputs[i].tokenAddress
+                );
+                if (
+                    recipe.inputs[i].tokenAction ==
+                    LibCrafting.INPUT_TOKEN_ACTION_TRANSFER
+                ) {
+                    token.transferFrom(
+                        msg.sender,
+                        address(this),
+                        recipe.inputs[i].amount
+                    );
+                } else if (
+                    recipe.inputs[i].tokenAction ==
+                    LibCrafting.INPUT_TOKEN_ACTION_BURN
+                ) {
+                    token.burnFrom(msg.sender, recipe.inputs[i].amount);
+                } else if (
+                    recipe.inputs[i].tokenAction ==
+                    LibCrafting.INPUT_TOKEN_ACTION_HOLD
+                ) {
+                    uint256 balance = token.balanceOf(msg.sender);
+                    if (balance < recipe.inputs[i].amount) {
+                        revert("User doesn't hold enough tokens for crafting");
+                    }
+                }
+            } else if (
+                recipe.inputs[i].tokenType == LibCrafting.ERC1155_TOKEN_TYPE
+            ) {
+                ERC1155Burnable token = ERC1155Burnable(
+                    recipe.inputs[i].tokenAddress
+                );
+                if (
+                    recipe.inputs[i].tokenAction ==
+                    LibCrafting.INPUT_TOKEN_ACTION_TRANSFER
+                ) {
+                    token.safeTransferFrom(
```

Comment:
Do we need to use `safeTransferFrom`? What is the downside of using just `transferFrom`?

Reply:
ERC1155 has only safe transfers:
https://eips.ethereum.org/EIPS/eip-1155#design-decision-safe-transfers-only

Reply:
fixed